Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,147 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "config.h"
+#include "GridMasonryLayout.h"
+
+#include "rendering/style/GridPositionsResolver.h"
+
+namespace WebCore {
+
+void GridMasonryLayout::performMasonryPlacement(HashMap<RenderBox*, GridArea>& firstTrackItems, Vector<RenderBox*> itemsWithDefiniteGridAxisPosition, Vector<RenderBox*> itemsWithIndefinitePosition, GridTrackSizingDirection masonryAxisDirection)
+{
+    m_masonryAxisDirection = masonryAxisDirection;
+    m_gridAxisDirection = m_masonryAxisDirection == ForRows ? ForColumns : ForRows;
+    m_masonryAxisGridGap = m_renderGrid.gridGap(m_masonryAxisDirection);
+    m_gridAxisTracksCount = RenderGrid::gridAxisTracksCountForMasonryLayout(m_renderGrid, m_gridAxisDirection);
+    m_runningPositions = Vector<LayoutUnit>(m_gridAxisTracksCount);
+
+    // 2.4 Masonry Layout Algorithm
+    addItemsToFirstTrack(firstTrackItems);
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto itemItr = firstTrackItems.begin(); itemItr != firstTrackItems.end(); ++itemItr) {
+        auto* item = itemItr->key;
+        auto gridArea = itemItr->value;
+
+        insertIntoGridAndLayoutItem(item, gridArea);
+
+        updateRunningPositions(item);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(Vector<RenderBox*> itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* childInGridOrder = m_grid.orderIterator().first(); childInGridOrder; childInGridOrder = m_grid.orderIterator().next()) {
+        if (!itemsWithDefiniteGridAxisPosition.contains(childInGridOrder))
```

Comment:
Searching in a vector may not be great for performance.

Why do items with definite position need to be iterated with `orderIterator()` if items with indefinite position or in first track aren't?

Reply:
Actually I think we can remove this logic for the definite position case and make it the same as the indefinite case. RenderGrid::collectMasonry items already goes through and stores the items into the appropriate vectors in the correct order.